docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md

```diff
@@ -0,0 +1,316 @@
+
+```solidity
+/// @notice This function is called by LayerZero when a cross-chain message is received
+    function _nonblockingLzReceive(
+        uint16 ,
+        bytes memory ,
+        uint64 _nonce,
+        bytes memory _payload
+ ) internal override {
+        // Decode the payload
+ (bytes memory toAddressBytes, uint256 amount, address tokenAddress, uint256 tokenId) = abi.decode(_payload, (bytes, uint256, address, uint256));
+
+        // Convert toAddressBytes to an address
+        address to;
+        assembly {
+ to := mload(add(toAddressBytes, 20))
+ }
+
+        // Increment the user's balance
+ balances[to] += amount;
+
+        // Emit events
+        emit CustomMessageReceived(amount, tokenAddress, tokenId, to);
+        emit EmitPayload(_payload);
+
+        // Notify the marketplace that the balance is updated (this could be another cross-chain message)
+        _notifyMarketplace(tokenAddress, tokenId, to, balances[to]);
+ }
+```
+
+The full contract code is as follows:
+
+```solidity reference
+https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/contracts/CrossChainAgent.sol
+```
+
+### NFTMarketPlaceV2.sol
+
+The `NFTMarketPlaceV2` contract is an updated version of the `NFTMarketPlace` contract that you deployed in [Part I](https://wiki.iota.org/isc/tutorials/cross-chain-nft-marketplace-part-1/). This version includes a new function `buyItemCrossChain` that is called by the `CrossChainAgent` contract on the ShimmerEVM Testnet to complete the purchase of an NFT.
+
+The function checks if the item is listed and available and if the user has enough balance to complete the purchase. It then transfers the NFT to the user by calling the NFT's proxy contract (the contract responsible for transferring the NFT cross-chain). The user's balance is updated, and the item's availability is set to false.
+
+```solidity
+    function buyItemCrossChain(
+        address nftAddress,
+        uint256 tokenId,
+        address payable buyer,
+        uint256 amount
+ ) external nonReentrant isListed(nftAddress, tokenId) {
+ Listing memory listedItem = s_listings[nftAddress][tokenId];
+
+        _validateAmount(amount, listedItem.price, nftAddress, tokenId);
+
+        delete s_listings[nftAddress][tokenId];
+
+        _approveNFT(listedItem.proxyContract, nftAddress, tokenId);
+
+        bytes memory adapterParams = _createAdapterParams(buyer);
+
+        uint256 fee = _estimateFee(listedItem.proxyContract, buyer, tokenId, adapterParams);
+
+        _sendNFT(listedItem, buyer, tokenId, adapterParams, fee);
+
+ s_proceeds[listedItem.seller] += amount;
+
+        emit ItemBought(msg.sender, nftAddress, tokenId, listedItem.price);
+ }
+```
+
+The full contract code is as follows:
+
+```solidity reference
+https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/contracts/NFTMarketPlaceV2.sol
+```
+
+
+## Scripts
```

Comment:
In the scripts section, we could also implement a command to verify the deployed contracts, wdyt?

Reply:
That would be a good addition yes, but I think it is out of scope for this tutorial. 
We would need the APIs for ShimmerEVM and BNBTestnet blockexplorers, and add some extra config to `hardhat.config.js`.
What would be more suitable is adding this part later on, WDYT?